[Streams] Child streams reordering on partitioning page

Kerry350 opened the pull request:
## Summary

Closes https://github.com/elastic/streams-program/issues/20.

Allows for drag and drop reordering of child streams on the partitioning page.

In the designs there are technically more stages (e.g. confirmation modals), however as these modals don't exist on the create / update steps either this implementation keeps the UX consistent with those steps also. 

![dnd](https://github.com/user-attachments/assets/40ec0b90-de9e-4ddb-8d7f-86fc2dfd3f7a)

## DnD note

There's an error in the console (dev only) when reordering as technically nested scroll containers aren't allowed (and between our resizable panels, regular panels etc there's several overflow contexts). However, this doesn't seem to actually break anything, so I think we're okay (the log won't appear in production).

![Screenshot 2025-01-13 at 13 16 54](https://github.com/user-attachments/assets/16d2c09d-40f4-47db-852e-9e46863bbc20)



<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

simianhacker commented:
Can we add an "Unsaved changes" prompt for the use case where a user re-orders the routing but forgets to click the save "Change routing" at the bottom of the page?

Here is how @tonyghiani implemented it: https://github.com/elastic/kibana/pull/204793/commits/b6fae34c99cad8d1c4b23c070634dc3230905cd9

flash1293 commented on x-pack/solutions/observability/plugins/streams_app/public/components/nested_view/index.tsx:
```diff
@@ -11,8 +11,18 @@ import { euiThemeVars } from '@kbn/ui-theme';
+}: {
+  children: React.ReactNode;
+  last?: boolean;
+  isBeingDragged: boolean;
```
Should probably be
```
  isBeingDragged?: boolean;
```

flash1293 reviewed (commented):
Looks mostly good to me - @simianhacker s comment makes sense to me.

Seems like we lost the margin between the entries, can we get them back?
<img width="711" alt="Screenshot 2025-01-10 at 15 13 25" src="https://github.com/user-attachments/assets/ee6aacc8-43d0-49ad-a489-cb4576d2229c" />

Just a nit I noticed (not introduced with your PR): When you reorder right away after adding a child, the success toast will cover the save button which is unfortunate. Not sure whether there is an easy way to avoid it though. Do you have an idea, @patpscal ?
<img width="1218" alt="Screenshot 2025-01-10 at 15 17 19" src="https://github.com/user-attachments/assets/208f9899-4ebe-4f6e-a7a9-06321cf0f912" />

Kerry350 commented:
@flash1293 Ready for another look. I've added the unsaved changes prompt, and fixed the margins. I haven't changed the toasts, it doesn't seem like there's an easy way to change the position (unless I'm missing something), but I think with the addition of the unsaved changes prompt this is *probably* okay.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/266490)
* Commit: 659db722463eb9409d47a6b0c2a02073b3ee112e

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b328fe89c9ac610fc324781a3574c6d6bdc1b0bb) | [after](https://github.com/elastic/kibana/commit/659db722463eb9409d47a6b0c2a02073b3ee112e) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 264.8KB | 267.2KB | +2.3KB |



### History
* :broken_heart: [Build #265749](https://buildkite.com/elastic/kibana-pull-request/builds/265749) failed 06b5d868a7e64ea488d57733e0ec1619d5f319a3

cc @Kerry350

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/266490","number":266490,"commit":"659db722463eb9409d47a6b0c2a02073b3ee112e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/265749","number":265749,"commit":"06b5d868a7e64ea488d57733e0ec1619d5f319a3"}],"number":266490}
buildkite-pr-comment-->

patpscal commented:
> In the designs there are technically more stages (e.g. confirmation modals), however as these modals don't exist on the create / update steps either this implementation keeps the UX consistent with those steps also.

@Kerry350 @LucaWintergerst Thank you! We included the confirmation modal to ensure users were fully aware of the impact of reordering on their data, but if that's no longer a requirement, we could use this approach! But we should use the same components as in Processors, so displaying a `bottom bar`.

<img width="962" alt="image" src="https://github.com/user-attachments/assets/4c78ac27-d2d8-468f-b1df-e3577d1f4310" />

patpscal commented:
Just a nit I noticed (not introduced with your PR): When you reorder right away after adding a child, the success toast will cover the save button which is unfortunate. Not sure whether there is an easy way to avoid it though. Do you have an idea, @patpscal ?

I see! Since the toast lingers a bit, it becomes a problem if changes are made quickly, although the toast has a dismiss button in the same spot, which helps to solve this issue. Would it make sense to dismiss the toast automatically once the user makes more changes (reorders again)?

flash1293 commented:
> Would it make sense to dismiss the toast automatically when/if the user makes more changes (reorders again) in the timeframe it's displayed?

That's a good idea!

flash1293 reviewed (approved):
LGTM, thanks for the changes!

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/12767641906

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/206570">](https://github.com/elastic/kibana/pull/206570)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
